### PR TITLE
Set log level run time

### PR DESCRIPTION
**Pull request by ap891843**
This story provides the following -

- a user is able to provide the settings
- settings can be changed on the fly
- The server can adjust the log level according to the settings

### CODE REVIEW

**Comment by ap891843** [comment]
Need modification.

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+   *
+   * @return Consumer#List#Object , currently supports List#String and List#JsonPrimitive
+   */
+  public static Consumer<List<Object>> updateLogLevel() {
+    return loglevel -> {
+      if (Objects.nonNull(loglevel)) {
+        Object localeNode = Iterables.getFirst(loglevel, "");
+        if (localeNode instanceof JsonPrimitive) {
+          updateLogLevel((JsonPrimitive) localeNode);
+        }
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java:52]: else if

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Update the log level for the application.
+ *
+ * <p>IMPORTANT: This class relies on the concrete implementations by logback. SLF4J don't expose
+ * any contract to update the log level on runtime.
+ *
+ * <p>In case of wrong input, log level defaults to {@link ch.qos.logback.classic.Level#ERROR}
+ */
+@UtilityClass
+public class LogLevel {
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java:36]: LogLevelUtils or LogLevelService

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Update loglevel for the entire application.
+   *
+   * @return Consumer#List#Object , currently supports List#String and List#JsonPrimitive
+   */
+  public static Consumer<List<Object>> updateLogLevel() {
+    return loglevel -> {
+      if (Objects.nonNull(loglevel)) {
+        Object localeNode = Iterables.getFirst(loglevel, "");
+        if (localeNode instanceof JsonPrimitive) {
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java:50]: Below you provide: logLevelNode.getAsString() which returns String, then why you need to check for "instanceof JsonPrimitive"

Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevel.java:50]: getAsString() is a method on JsonPrimitive, that's why casting it to JsonPrimitive upon instanceof check.

`private static void updateLogLevel(JsonPrimitive logLevelNode)`

accepts JsonPrimitive

**Comment by ap891843** [comment]
Needs improvement.

**Comment by ap891843** [comment]
Now it's ready 🚶‍♂️

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevelUtils.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Update loglevel for the entire application.
+   *
+   * @return Consumer#List#Object , currently supports List#String and List#JsonPrimitive
+   * @throws IllegalArgumentException currently supports only List#String and List#JsonPrimitive
+   */
+  public static Consumer<List<Object>> updateLogLevel() {
+    return loglevel -> {
+      if (Objects.nonNull(loglevel)) {
+        Object localeNode = Iterables.getFirst(loglevel, "");
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LogLevelUtils.java:51]: `logNode`?

File: server/src/test/java/com/broadcom/lsp/cobol/core/messages/LogLevelUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class LogLevelUtilsTest {
```
Comment (temanbrcom) [inline server/src/test/java/com/broadcom/lsp/cobol/core/messages/LogLevelUtilsTest.java:28]: Missing javadoc

**Review by temanbrcom — CHANGES_REQUESTED** [review]
The log level settings are not exposed to the settings UI. You should add it in `configuration` section of `properties.json` file

**Comment by temanbrcom** [comment]
I would kindly ask to rebase and squash the actual commits and add a "feat: " prefix.

**Comment by temanbrcom** [comment]
> The log level settings are not exposed to the settings UI. You should add it in `configuration` section of `properties.json` file

Could you, please, also check if we may specify the predefined values for the log level in order to let the user choose one of them

**Comment by ap891843** [comment]
> > The log level settings are not exposed to the settings UI. You should add it in `configuration` section of `properties.json` file
> 
> Could you, please, also check if we may specify the predefined values for the log level in order to let the user choose one of them

That would be great. Yes, I will check that.

**Comment by ap891843** [comment]
Logging Support -  che Testing:

image

**Comment by ap891843** [comment]
Che test: locale support

image
image

File: clients/cobol-lsp-vscode-extension/package.json [hunk 6]
```diff
@@ -108,6 +108,32 @@
...
+                        "TRACE",
+                        "ALL"
+                    ],
+                    "enumDescriptions": [
+                        "ERROR level logs error events which may or not be fatal to the application",
+                        "DEBUG level logs informational events of very low importance",
+                        "INFO level logs informational messages highlighting overall progress of the application",
+                        "WARN level logs potentially harmful situations",
+                        "TRACE level logs informational events of very low importance",
+                        "ALL level is used to turn on all logging"
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/package.json:129]: The top-down order should be ERROR-WARN-INFO-DEBUG-TRACE-ALL

File: clients/cobol-lsp-vscode-extension/package.json [hunk 7]
```diff
@@ -108,6 +108,32 @@
...
+                    ],
+                    "description": "log level for COBOL LSP server."
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/package.json:131]:
```suggestion
                    "description": "The logging level for COBOL LS backend"
```

**Review by temanbrcom — CHANGES_REQUESTED** [review]

File: clients/cobol-lsp-vscode-extension/package.json [hunk 8]
```diff
@@ -108,6 +108,32 @@
...
+                },
+                "broadcom-cobol-lsp.locale": {
+                    "type": "string",
+                    "default": "en",
+                    "description": "locale for the COBOL LSP server."
```
[the remark this answers is not in the document]
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/package.json]: Done.

**Review by temanbrcom — APPROVED** [review]
🐙

File: clients/cobol-lsp-vscode-extension/package.json [hunk 9]
```diff
@@ -108,6 +108,27 @@
...
+                    ],
+                    "enumDescriptions": [
+                        "ERROR level logs error events which may or not be fatal to the application",
+                        "WARN level logs potentially harmful situations",
+                        "INFO level logs informational messages highlighting overall progress of the application",
+                        "DEBUG level logs informational events of very low importance",
+                        "TRACE level logs informational events of very low importance",
+                        "ALL level is used to turn on all logging"
+                    ],
+                    "description": "The logging level for COBOL LS backend"
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/package.json:131]: LSP?

**Review by asatklichov — APPROVED** [review]
🔢

File: clients/cobol-lsp-vscode-extension/package.json (hunk 9 above)
Comment (temanbrcom) [reply clients/cobol-lsp-vscode-extension/package.json:131]: COBOL Language Support

**Pull request merged**